### PR TITLE
Experimental gather prefetch

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Added prefetching only to the `cudf::gather` API. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/strings/detail/gather.cuh [hunk 1]
```diff
@@ -310,6 +322,9 @@ std::unique_ptr<cudf::column> gather(strings_column_view const& strings,
     sizes_itr, sizes_itr + output_count, stream, mr);
 
   // build chars column
+  cudf::experimental::prefetch::detail::prefetch("prefetch", out_offsets_column->view());
+  cudf::experimental::prefetch::detail::prefetch(
+    "prefetch", strings.chars_begin(stream), strings.chars_size(stream));
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/gather.cuh:327]: @vyasr I think this is a key one that will not be handled by the containers.

File: cpp/include/cudf/strings/detail/strings_children.cuh [hunk 2]
```diff
@@ -187,6 +193,7 @@ auto make_strings_children(SizeAndExecuteFunction size_and_exec_fn,
   // Now build the chars column
   rmm::device_uvector<char> chars(bytes, stream, mr);
   size_and_exec_fn.d_chars = chars.data();
+  cudf::experimental::prefetch::detail::prefetch("prefetch", chars);
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/strings_children.cuh:196]: @vyas This one replaces the subclassed uvector.

File: cpp/include/cudf/strings/detail/gather.cuh [hunk 3]
```diff
@@ -231,6 +233,7 @@ rmm::device_uvector<char> gather_chars(StringIterator strings_begin,
 
   auto chars_data = rmm::device_uvector<char>(chars_bytes, stream, mr);
   auto d_chars    = chars_data.data();
+  cudf::experimental::prefetch::detail::prefetch("prefetch", chars_data);
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/gather.cuh:236]: @vyas This replaces the subclassed uvector too

File: cpp/include/cudf/strings/detail/gather.cuh [hunk 4]
```diff
@@ -298,6 +301,15 @@ std::unique_ptr<cudf::column> gather(strings_column_view const& strings,
...
     strings.offset());
 
+  if (!strings.is_empty()) {
+    auto offsets =
+      cudf::slice(strings.offsets(), {strings.offset(), strings.offset() + strings.size()}, stream)
+        .front();
+    cudf::experimental::prefetch::detail::prefetch("prefetch", offsets);
+    cudf::experimental::prefetch::detail::prefetch(
+      "prefetch", strings.null_mask(), cudf::bitmask_allocation_size_bytes(strings.size()));
+  }
```
Comment (vyasr) [inline cpp/include/cudf/strings/detail/gather.cuh:311]: I looked through the PR and I independently came to the same conclusions as your comments regarding which prefetch calls I needed to include in my PR, except for this bit. Is this changeset significant?

Comment (davidwendt) [reply cpp/include/cudf/strings/detail/gather.cuh:311]: This handles prefetching the strings offsets honoring a sliced input column.
The `column_view` overrides in #16020 should prefetch the offsets automatically I believe.
Though the #16020 override of `column_view::head()` does not account for a sliced `column_view` and so may fetch more data than actually used by the kernel -- a potential issue if the caller is using `slice/split` to partition workload. 
The `prefetch(key,column_view)` overload in this PR handles prefetching sliced columns.

**Comment by vyasr** [comment]
Closing in favor of #16020.

**Pull request closed without merging**